runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/net/IpAddr.kt

```diff
@@ -0,0 +1,416 @@
+                        segments.forEachIndexed { idx, segment ->
+                            if (segment == 0.toUShort()) {
+                                if (curr.len == 0) curr.start = idx
+
+                                curr.len += 1
+
+                                if (curr.len > longest.len) longest = curr
+                            } else {
+                                curr = Span()
+                            }
+                        }
+                        longest
+                    }
+
+                    if (zeroes.len > 1) {
+                        formatSegments(0 until zeroes.start)
+                        append("::")
+                        formatSegments((zeroes.start + zeroes.len) until segments.size)
+                    } else {
+                        formatSegments(segments.indices)
+                    }
+                }
+            }
+
+            if (zoneId != null) {
+                "$formatted%$zoneId"
+            } else {
+                formatted
+            }
+        }
+
+        private fun StringBuilder.formatSegments(range: IntRange) {
+            if (range.first >= segments.size || range.isEmpty()) return
+            append(segments[range.first].toString(16))
+            val tail = IntRange(range.first + 1, range.last)
+            for (i in tail) {
+                append(':')
+                append(segments[i].toString(16))
+            }
+        }
+
+        /**
+         * Returns true if this is the loopback address (`::1`), as defined in
+         * [RFC 4291 section 2.5.3.](https://www.rfc-editor.org/rfc/rfc4291#section-2.5.3)
+         */
+        override val isLoopBack: Boolean
+            get() = this == LOCALHOST
+
+        /**
+         * Returns true if this is the "any" address (`::`)
+         */
+        public val isUnspecified: Boolean
+            get() = this == UNSPECIFIED
+
+        /**
+         * Returns the multicast scope of the address if it is a multicast address
+         */
+        public val multicastScope: Ipv6MulticastScope?
+            get() = when (isMulticast) {
+                false -> null
+                true -> when ((segments[0] and 0x000fu).toUInt()) {
+                    1u -> Ipv6MulticastScope.InterfaceLocal
+                    2u -> Ipv6MulticastScope.LinkLocal
+                    3u -> Ipv6MulticastScope.RealmLocal
+                    4u -> Ipv6MulticastScope.AdminLocal
+                    5u -> Ipv6MulticastScope.SiteLocal
+                    8u -> Ipv6MulticastScope.OrganizationLocal
+                    14u -> Ipv6MulticastScope.Global
+                    else -> null
+                }
+            }
+
+        /**
+         * Returns true if this is a multicast address as defined by
+         * [RFC 4291 2.7](https://www.rfc-editor.org/rfc/rfc4291#section-2.7).
+         */
+        public val isMulticast: Boolean
+            get() = segments[0] and 0xff00u == 0xff00u.toUShort()
```

Comment:
Yeah I think there is definitely a bit more to be exposed. My aim was to get a fairly reasonable representation of an API for IP addresses regardless of what we consume. I can understand going the other way though and exposing a more minimal surface area until need arises. What is your preference and why? Also if you're preference is a minimal subset what subset do you propose?

Reply:
My concern is support burden. Exposing these properties means we must support them and even keep them up to date if more multicast scopes are introduced. We've chosen in other places to limit our implementations and APIs to only the features our SDK truly needs (e.g., deserialization) instead of providing general-use library functionality.

My preference for a minimal subset would be only the things which we need now or know we're going to need shortly for upcoming implementations. Other than that I'd assume we're not going to need it and could add it later should that change.

Reply:
This likely mostly comes down to removing the multicast flags then I think, the rest of this seems like the bare bones API I'd want to define (IPv4, IPv6, parsing, toString, isLoopBack, and isUnspecified).